src/diffusers/models/autoencoders/dc_ae.py

```diff
@@ -0,0 +1,877 @@
+    elif block_type == "InterpolateConv":
+        block = Upsample2D(channels=in_channels, use_conv=True, out_channels=out_channels)
+    else:
+        raise ValueError(f"block_type {block_type} is not supported for upsampling")
+    if shortcut is None:
+        pass
+    elif shortcut == "duplicating":
+        shortcut_block = ChannelDuplicatingPixelUnshuffleUpsample2D(
+            in_channels=in_channels, out_channels=out_channels, factor=2
+        )
+        block = ResidualBlock(block, shortcut_block)
+    else:
+        raise ValueError(f"shortcut {shortcut} is not supported for upsample")
+    return block
+
+
+def build_encoder_project_in_block(in_channels: int, out_channels: int, factor: int, downsample_block_type: str):
+    if factor == 1:
+        block = nn.Conv2d(
+            in_channels=in_channels,
+            out_channels=out_channels,
+            kernel_size=3,
+            padding=1,
+        )
+    elif factor == 2:
+        block = build_downsample_block(
+            block_type=downsample_block_type, in_channels=in_channels, out_channels=out_channels, shortcut=None
+        )
+    else:
+        raise ValueError(f"downsample factor {factor} is not supported for encoder project in")
+    return block
+
+
+def build_encoder_project_out_block(
+    in_channels: int, out_channels: int, norm: Optional[str], act: Optional[str], shortcut: Optional[str]
+):
+    block = OpSequential(
+        [
+            build_norm(norm),
+            get_activation(act) if act is not None else None,
+            ConvLayer(
+                in_channels=in_channels,
+                out_channels=out_channels,
+                kernel_size=3,
+                stride=1,
+                use_bias=True,
+                norm=None,
+                act_func=None,
+            ),
+        ]
+    )
+    if shortcut is None:
+        pass
+    elif shortcut == "averaging":
+        shortcut_block = PixelUnshuffleChannelAveragingDownsample2D(
+            in_channels=in_channels, out_channels=out_channels, factor=1
+        )
+        block = ResidualBlock(block, shortcut_block)
+    else:
+        raise ValueError(f"shortcut {shortcut} is not supported for encoder project out")
+    return block
+
+
+def build_decoder_project_in_block(in_channels: int, out_channels: int, shortcut: Optional[str]):
+    block = ConvLayer(
+        in_channels=in_channels,
+        out_channels=out_channels,
+        kernel_size=3,
+        stride=1,
+        use_bias=True,
+        norm=None,
+        act_func=None,
+    )
+    if shortcut is None:
+        pass
+    elif shortcut == "duplicating":
+        shortcut_block = ChannelDuplicatingPixelUnshuffleUpsample2D(
+            in_channels=in_channels, out_channels=out_channels, factor=1
+        )
+        block = ResidualBlock(block, shortcut_block)
+    else:
+        raise ValueError(f"shortcut {shortcut} is not supported for decoder project in")
+    return block
+
+
+def build_decoder_project_out_block(
```

Comment:
let's try to remove this and make the layers directly inside the `__init__` of `Encoder` instead
also, we do not need to consider configuration that's not used for the checkpoint you are releasing this time, e.g. I think we only need to include logic for `factor == 1` or `factor == 2` depends on the configuration used (but not both)

example for how to do this directly inside `__init__`

```python
class Encoder():

def __init__(self, ...):
       ....
        self.act_out = get_activation("silu")
        self.proj_out = ConvPixelShuffleUpsample2D(...)
```

Reply:
If we know the configurations of the checkpoints being released I think we don't need to define a function to retrieve the configs as Yiyi mentioned and make layers initialized directly inside.

Reply:
Sure. I'll do that.

Reply:
done